storage: Confirmation before erasing non-empty filesystems

mvollmer opened the pull request:
But only in Anaconda mode.

Demo: https://youtu.be/EYU2jzDI0gk

------

### Storage: Cockpit asks for extra confirmation before deleting files in Anaconda mode

When using Cockpit to configure storage manually during installation with Anaconda, one normally only operates on empty filesystems. You might create a couple of partitions with filesystems and then change your mind and delete those partitions again, maybe to create them in a slightly different way. These are all pretty harmless operations since no actual data gets deleted.

However, there might also be filesystems present that have not been created during installation. They might contain other operating systems and old, valueable files. In order to protect against deleting these by accident, Cockpit now checks whether a filesystem contains actual files before deleting it, and asks for extra confirmation if it does.

![image](https://github.com/cockpit-project/cockpit/assets/3228183/dbeadcf7-b3f1-49ff-a6ea-e483b8ced09e)

cockpituous commented on pkg/storaged/dialog.jsx:
```diff
@@ -1247,6 +1296,24 @@ export const TeardownMessage = (usage, expect_single_unmount) => {
+            }
+            { have_data && usage.length > 1 &&
+              <>
+                  <br/>
+                  <ExtraConfirmation dlg={usage.dlg}
+                                     title={_("Some filesystems are not empty.")}
+                                     text={_("Some of the devices that you are about to erase contain files. Please confirm that you really want to erase them.")}
+                                     confirm_text={_("Yes, erase all files on these devices")} />
```
These 5 added lines are not executed by any test.

cockpituous commented on pkg/storaged/dialog.jsx:
```diff
@@ -1247,6 +1293,24 @@ export const TeardownMessage = (usage, expect_single_unmount) => {
+            }
+            { have_data && usage.length > 1 &&
+              <>
+                  <br/>
+                  <ExtraConfirmation dlg={usage.dlg}
+                                     title={_("Some filesystems are not empty.")}
+                                     text={_("Some of the devices that you are about to erase contain files. Please confirm that you really want to erase them.")}
+                                     confirm_text={_("Yes, erase all files on these devices")} />
```
These 5 added lines are not executed by any test.

cockpituous commented on pkg/storaged/block/resize.jsx:
```diff
@@ -647,7 +647,7 @@ export function shrink_dialog(client, lvol_or_part, info, to_fit) {
         },
         Inits: [
-            init_active_usage_processes(client, usage),
+            init_teardown_usage(client, usage),
```
This added line is not executed by any test.

cockpituous commented on pkg/storaged/btrfs/subvolume.jsx:
```diff
@@ -204,6 +204,7 @@ function subvolume_delete(volume, subvol, mount_point_in_parent, card) {
     const paths_to_delete = [];
     const usage = [];
 
+    usage.Teardown = true;
```
This added line is not executed by any test.

cockpituous commented on pkg/storaged/btrfs/subvolume.jsx:
```diff
@@ -241,7 +242,7 @@
         },
         Inits: [
-            init_active_usage_processes(client, usage)
+            init_teardown_usage(client, usage)
```
This added line is not executed by any test.

cockpituous commented on pkg/storaged/dialog.jsx:
```diff
@@ -415,8 +423,10 @@ export const dialog_open = (def) => {
-                                                      errors && errors.toString() != "[object Object]"),
+                disabled: (running_promise != null ||
+                           (def.Action.disable_on_error &&
+                            errors && errors.toString() != "[object Object]") ||
+                           (need_confirmation && !confirmed)),
```
These 2 added lines are not executed by any test.

cockpituous commented on pkg/storaged/dialog.jsx:
```diff
@@ -537,6 +547,17 @@
             update();
         },
 
+        need_confirmation: () => {
+            need_confirmation = true;
+            confirmed = false;
+            update_footer();
```
These 4 added lines are not executed by any test.

cockpituous commented on pkg/storaged/dialog.jsx:
```diff
@@ -537,6 +547,17 @@
+            update_footer();
+        },
+
+        set_confirmed: (flag) => {
+            confirmed = flag;
+            update_footer();
```
These 3 added lines are not executed by any test.

cockpituous commented on pkg/storaged/dialog.jsx:
```diff
@@ -1206,14 +1227,33 @@
     return name;
 };
 
+const ExtraConfirmation = ({ title, text, confirm_text, dlg }) => {
+    const [confirmed, setConfirmed] = useState(false);
```
These 2 added lines are not executed by any test.

cockpituous commented on pkg/storaged/dialog.jsx:
```diff
@@ -1206,14 +1227,33 @@
+const ExtraConfirmation = ({ title, text, confirm_text, dlg }) => {
+    const [confirmed, setConfirmed] = useState(false);
+
+    return (
+        <Alert variant="danger" isInline
+               title={title}>
+            {text}
+            <Checkbox isChecked={confirmed}
+                      id="dialog-confirm"
+                      label={confirm_text}
+                      onChange={(_, val) => {
+                          console.log("C", val);
+                          setConfirmed(val);
+                          dlg.set_confirmed(val);
```
These 11 added lines are not executed by any test.

cockpituous commented on pkg/storaged/dialog.jsx:
```diff
@@ -1206,14 +1227,33 @@
+                          setConfirmed(val);
+                          dlg.set_confirmed(val);
+                      }} />
+        </Alert>);
```
This added line is not executed by any test.

cockpituous commented on pkg/storaged/dialog.jsx:
```diff
@@ -1223,12 +1263,22 @@
                     location = _("(Not part of target)");
             }
+            if (use.data_warning)
+                have_data = true;
```
This added line is not executed by any test.

cockpituous commented on pkg/storaged/dialog.jsx:
```diff
@@ -1223,12 +1263,22 @@
+                        title: <>
+                            <UsersPopover users={use.users || []} />
+                            { use.data_warning &&
+                            <span>
+                                <ExclamationTriangleIcon className="ct-icon-exclamation-triangle" /> { "\n" }
+                                { use.data_warning }
+                            </span>
```
These 4 added lines are not executed by any test.

cockpituous commented on pkg/storaged/dialog.jsx:
```diff
@@ -1247,6 +1297,15 @@
                           rows={rows} />
+            { have_data &&
+            <>
+                <br />
+                <ExtraConfirmation dlg={usage.dlg}
+                                     title={_("Important data might be deleted.")}
+                                     text={_("The device that you are about to erase or delete might contain important data. Please confirm that you really want to delete or erase this device.")}
+                                     confirm_text={_("Yes, delete my data.")} />
```
These 5 added lines are not executed by any test.

cockpituous commented on pkg/storaged/dialog.jsx:
```diff
@@ -1268,24 +1327,45 @@
+                                                           [decode_filename(u.block.PreferredDevice)],
+                                                           { superuser: true, err: "message" });
+                        if (empty.trim() != "yes")
+                            u.data_warning = _("Filesystem is not empty");
+                    } else if (u.block.IdUsage == "crypto" && !client.blocks_cleartext[u.block.path]) {
+                        u.data_warning = _("Locked encrypted device might contain data");
+                    } else if (!client.blocks_ptable[u.block.path]
```
These 4 added lines are not executed by any test.

cockpituous commented on pkg/storaged/dialog.jsx:
```diff
@@ -1268,24 +1327,45 @@
+                        u.data_warning = _("Locked encrypted device might contain data");
+                    } else if (!client.blocks_ptable[u.block.path]
+                               && u.block.IdUsage != "raid"
+                               && !is_available_block(client, u.block)) {
+                        u.data_warning = _("Device contains unrecognized data");
```
These 2 added lines are not executed by any test.

cockpituous commented on pkg/storaged/dialog.jsx:
```diff
@@ -1268,24 +1327,45 @@
+                        u.data_warning = _("Device contains unrecognized data");
+                    }
+                    if (u.data_warning)
+                        have_data = true;
```
This added line is not executed by any test.

cockpituous commented on pkg/storaged/dialog.jsx:
```diff
@@ -1268,24 +1327,45 @@
+            }
+
+            usage.dlg = dlg;
+            if (have_data) {
+                usage.Teardown = true;
+                dlg.need_confirmation();
```
These 3 added lines are not executed by any test.

cockpituous commented on pkg/storaged/legacy-vdo/legacy-vdo.jsx:
```diff
@@ -68,7 +68,7 @@ export function make_legacy_vdo_page(parent, vdo, backing_block, next_card) {
                 },
                 Inits: [
-                    init_active_usage_processes(client, usage)
+                    init_teardown_usage(client, usage)
```
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

cockpituous commented on pkg/storaged/legacy-vdo/legacy-vdo.jsx:
```diff
@@ -130,7 +130,7 @@
             },
             Inits: [
-                init_active_usage_processes(client, usage)
+                init_teardown_usage(client, usage)
```
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

cockpituous commented on pkg/storaged/lvm2/block-logical-volume.jsx:
```diff
@@ -85,7 +85,7 @@ export function lvol_delete(lvol, card) {
         },
         Inits: [
-            init_active_usage_processes(client, usage)
+            init_teardown_usage(client, usage)
```
This added line is not executed by any test.

cockpituous commented on pkg/storaged/lvm2/block-logical-volume.jsx:
```diff
@@ -166,7 +166,7 @@
         },
         Inits: [
-            init_active_usage_processes(client, usage)
+            init_teardown_usage(client, usage)
```
This added line is not executed by any test.

cockpituous commented on pkg/storaged/lvm2/volume-group.jsx:
```diff
@@ -104,7 +104,7 @@ function vgroup_delete(client, vgroup, card) {
         },
         Inits: [
-            init_active_usage_processes(client, usage)
+            init_teardown_usage(client, usage)
```
This added line is not executed by any test.

cockpituous commented on pkg/storaged/lvm2/volume-group.jsx:
```diff
@@ -328,8 +328,10 @@
 
-        let usage = get_active_usage(client, vgroup.path, _("delete"));
-        usage = usage.filter(u => u.block && is_partial_linear_lvol(u.block));
+        const all_usage = get_active_usage(client, vgroup.path, _("delete"));
+        const usage = all_usage.filter(u => u.block && is_partial_linear_lvol(u.block));
+        usage.Blocking = all_usage.Blocking;
+        usage.Teardown = all_usage.Teardown;
```
These 4 added lines are not executed by any test.

cockpituous commented on pkg/storaged/lvm2/volume-group.jsx:
```diff
@@ -359,7 +361,7 @@
             },
             Inits: [
-                init_active_usage_processes(client, usage)
+                init_teardown_usage(client, usage)
```
This added line is not executed by any test.

cockpituous commented on pkg/storaged/mdraid/mdraid.jsx:
```diff
@@ -81,7 +81,7 @@ function mdraid_stop(mdraid) {
             },
             Inits: [
-                init_active_usage_processes(client, usage)
+                init_teardown_usage(client, usage)
```
This added line is not executed by any test.

cockpituous commented on pkg/storaged/mdraid/mdraid.jsx:
```diff
@@ -132,7 +132,7 @@
         },
         Inits: [
-            init_active_usage_processes(client, usage)
+            init_teardown_usage(client, usage)
```
This added line is not executed by any test.

cockpituous commented on pkg/storaged/stratis/pool.jsx:
```diff
@@ -154,7 +154,7 @@ function delete_pool(pool, card) {
         },
         Inits: [
-            init_active_usage_processes(client, usage)
+            init_teardown_usage(client, usage)
```
This added line is not executed by any test.

garrett commented:
Here's some revised mockups that streamline things:

![Initilize disk](https://github.com/cockpit-project/cockpit/assets/10246/1542205c-b6cf-4c46-99ff-6bd33d949c0c)

1. Don't show two or three options in a dropdown (that should really be for 5 or more items only, or a dynamic amount that _can_ be 5 or more); I changed it to radios instead
2. I renamed it to "modern" and "legacy" with GPT and MBR specified in parenthesis
3. The 2 TB limit is specified on the legacy MBR option, as it's a limitation of the old method, not a feature of the new
4. The warning is not a warning (as it is not the result of an action), but a helper text
5. The table is without much else, so it's _just_ the possibly affected partitions with data
6. The table contains the filesystem label as well as the size (used & total)
7. I dropped the /dev/ to be more consistent with the rest of Cockpit and Anaconda (as /dev/ is redundant for devices, as every device is in /dev/)
8. Use PatternFly suggestions (for the heading, for example)
9. Remove overwrite with zeros option, as per our few discussions about it, as [it's problematic for all kinds of reasons](https://wiki.archlinux.org/title/Securely_wipe_disk) (false sense of security, isn't consistently supported, physically damages drives, can destroy drives in some cases, and the actual solution for ensuring sensitive data is not on the drive is to have it encrypted from the start).

This is all assuming we detect data like we talked about yesterday. That is, show the warning only if:

1. It's not a partition created in this Anaconda and/or Cockpit session
2. It has data
   - Ideally without mounting the partition
   - We could check the used size; if it's over a certain small threshold, we know there are files
   - If it's encrypted

Examples:
- If someone created an encrypted partition within the session, it wouldn't get flagged as containing data, as it was created in the session.
- If someone created a partition with another tool ahead of time, but it does not contain data (as in, the used size is barely anything), then it wouldn't be flagged as having data.

This would let Cockpit not get in the way when someone is adding and removing partitions, but it also would protect someone from accidentally losing data.

We should use this general pattern for initializing the disk, formatting a partition, and anything else that would probably cause someone to lose data on a partition like this. Although we should rework the warning area a little with regard to the text (and probably table too).

Open questions:
- Should we not even include an option of GPT vs. MBR for disks that are 2TB or more?
- Should we really include options for GPT vs. MBR at all? (Is it actually important in 2024?)

mvollmer commented:
> Here's some revised mockups that streamline things:

Thanks!

I think we should split the parts out that are general improvements to the "Initialize disk" dialog. They might even happen as part of #19882.

mvollmer commented:
I think "a partition on this disk" is too specific, there are other cases like LVM2 logical volumes. What about "Important data might be deleted" as the header?

mvollmer commented:
@garrett, looks like this now: 

![image](https://github.com/cockpit-project/cockpit/assets/3228183/13259c80-6044-4d83-89a5-924f3fcefd6f)

The text of the "important data" section pretty small. This is the result of using `HelperText` and a compact `Table`...

cockpituous commented on pkg/storaged/dialog.jsx:
```diff
@@ -1268,24 +1350,50 @@ export function teardown_danger_message(usage, expect_single_unmount) {
+                        }
+                    } else if (u.block.IdUsage == "crypto" && !client.blocks_cleartext[u.block.path]) {
+                        u.data_warning = _("Locked encrypted device might contain data");
+                    } else if (!client.blocks_ptable[u.block.path] &&
```
This added line is not executed by any test.

cockpituous commented on pkg/storaged/dialog.jsx:
```diff
@@ -1268,24 +1350,50 @@ export function teardown_danger_message(usage, expect_single_unmount) {
+                        u.data_warning = _("Locked encrypted device might contain data");
+                    } else if (!client.blocks_ptable[u.block.path] &&
+                               u.block.IdUsage != "raid" &&
+                               !is_available_block(client, u.block)) {
+                        u.data_warning = _("Device contains unrecognized data");
```
These 2 added lines are not executed by any test.

cockpituous commented on pkg/storaged/dialog.jsx:
```diff
@@ -1268,24 +1350,50 @@ export function teardown_danger_message(usage, expect_single_unmount) {
+            if (have_data) {
+                usage.Teardown = true;
+                dlg.need_confirmation(_("I confirm I want to lose this data forever"));
+            } else if (client.in_anaconda_mode() && !expect_single_unmount) {
```
This added line is not executed by any test.

cockpituous commented on pkg/storaged/legacy-vdo/legacy-vdo.jsx:
```diff
@@ -130,7 +130,7 @@ export function make_legacy_vdo_page(parent, vdo, backing_block, next_card) {
             },
             Inits: [
-                init_active_usage_processes(client, usage)
+                init_teardown_usage(client, usage)
```
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

mvollmer commented:
>     1. Don't show two or three options in a dropdown (that should really be for 5 or more items only, or a dynamic amount that _can_ be 5 or more); I changed it to radios instead
> 
>     2. I renamed it to "modern" and "legacy" with GPT and MBR specified in parenthesis
> 
>     3. The 2 TB limit is specified on the legacy MBR option, as it's a limitation of the old method, not a feature of the new
>
>     9. Remove overwrite with zeros option, as per our few discussions about it, as [it's problematic for all kinds of reasons](https://wiki.archlinux.org/title/Securely_wipe_disk) (false sense of security, isn't consistently supported, physically damages drives, can destroy drives in some cases, and the actual solution for ensuring sensitive data is not on the drive is to have it encrypted from the start).
 
This has been implemented in #19882.

>     4. The warning is not a warning (as it is not the result of an action), but a helper text
> 
>     5. The table is without much else, so it's _just_ the possibly affected partitions with data
> 
>     6. The table contains the filesystem label as well as the size (used & total)
> 
>     7. I dropped the /dev/ to be more consistent with the rest of Cockpit and Anaconda (as /dev/ is redundant for devices, as every device is in /dev/)

This has been partly implemented here (for the "important data" table), and will be completed in #19882.
 
>     8. Use PatternFly suggestions (for the heading, for example)

I don't understand this, can you elabotrate?

cockpituous commented on pkg/storaged/dialog.jsx:
```diff
@@ -1268,24 +1339,50 @@ export function teardown_danger_message(usage, expect_single_unmount) {
+            if (have_data) {
+                usage.Teardown = true;
+                dlg.need_confirmation(_("I confirm I want to lose this data forever"));
+            } else if (client.in_anaconda_mode() && !expect_single_unmount) {
```
This added line is not executed by any test.

mvollmer commented:
>     * Should we not even include an option of GPT vs. MBR for disks that are 2TB or more?

I did this in #19882.

mvollmer commented:
> This is all assuming we detect data like we talked about yesterday. That is, show the warning only if:
> 
>  1. It's not a partition created in this Anaconda and/or Cockpit session

I am reluctant to implement this. We would have to identify all places in Cockpit where things get created. That's of course possible, but I hope we can find a more straighforward rule to determine whether to show the warning or not, a rule that doesn't require us to gather information as the user is using Cockpit. That would be more robust.

Right now the rule is:

  - If it is a filesystem of a recognized type (xfs, ext2, ext3, ext4, btrfs, vfat, or ntfs), we check whether it has anything in it that is not a directory. This might mount the filesystem temporarily.
  - If it is a encrypted device that is locked, we say "Locked encrypted device might contain data" and warn about it regardless of what is in it.
  - If it is something not recognized by Cockpit in general or a filesystem not in the list above, we say "Device contains unrecognized data"

My hope is that encrypted devices that are created by Cockpit will remain open during the installation process.

The idea behind the list of recognized filesystem types is to avoid trying to mount unusual filesystems that might be supported by the kernel, but are maybe buggy.

cockpituous commented on pkg/storaged/dialog.jsx:
```diff
@@ -1268,24 +1339,51 @@ export function teardown_danger_message(usage, expect_single_unmount) {
+            if (have_data) {
+                usage.Teardown = true;
+                dlg.need_confirmation(_("I confirm I want to lose this data forever"));
+            } else if (client.in_anaconda_mode() && !expect_single_unmount) {
```
This added line is not executed by any test.

jelly commented on pkg/storaged/dialog.jsx:
```diff
@@ -1269,24 +1339,51 @@ export function teardown_danger_message(usage, expect_single_unmount) {
+                }
+                if (client.in_anaconda_mode() && !expect_single_unmount && u.block) {
+                    if (u.block.IdUsage == "filesystem" &&
+                        ["xfs", "ext2", "ext3", "ext4", "btrfs", "vfat", "ntfs"].indexOf(u.block.IdType) >= 0) {
```
Are these all the filesystems udisks supports? If so Cockpit should already have a list of them somewhere right? Or can't we get this from udisks? Like http://storaged.org/doc/udisks2-api/latest/gdbus-org.freedesktop.UDisks2.Manager.html#gdbus-property-org-freedesktop-UDisks2-Manager.SupportedFilesystems

mvollmer replied:
> Are these all the filesystems udisks supports?

No, the idea is to only mount a small number of well-known filesystems, instead of just throwing everything that we find at the kernel.

jelly commented on pkg/storaged/dialog.jsx:
```diff
@@ -1269,24 +1339,51 @@ export function teardown_danger_message(usage, expect_single_unmount) {
+
+            if (have_data) {
+                usage.Teardown = true;
+                dlg.need_confirmation(_("I confirm I want to lose this data forever"));
```
Sounds a bit dramatic :laughing:

mvollmer replied:
It's a dramatic situation! :-)

jelly commented on pkg/storaged/fsys-is-empty.sh:
```diff
@@ -0,0 +1,29 @@
+dev=$1
+
+need_unmount=""
+mp=$(findmnt -no TARGET "$dev" | cat)
```
That cat seems unnecessary:

```
[jelle@t14s][~/projects/cockpit/main]%x=$(findmnt -no TARGET /dev/nvme0n1p1)
[jelle@t14s][~/projects/cockpit/main]%echo $x
/boot
```

mvollmer replied:
It protects against the exit code of `findmnt`. When $dev is not mounted, findmnt exits with code 1, which would stop the script because of `set -e`.  In a pipeline, the last sub-command determines the exit code of the while pipeline, and `cat` always exits with 0.

jelly commented on pkg/storaged/fsys-is-empty.sh:
```diff
@@ -0,0 +1,29 @@
+# A filesystem is empty if it only has directories in it.
+
+first=$(find "$mp" -not -type d | head -1)
+info=$(stat -f -c '{ "unit": %S, "free": %f, "total": %b }' "$mp")
```
Neither of these commands is allowed to fail as then we will exit and never unmount if needed. So maybe we should use a trap for that.

(I can only get find to fail when not running as root for example but maybe it can fail on broken symlinks or pipes as root?)

mvollmer replied:
> So maybe we should use a trap for that.

Good point!

jelly commented on test/verify/check-storage-anaconda:
```diff
@@ -549,6 +544,86 @@ class TestStorageAnaconda(storagelib.StorageCase):
+        m.execute(f"echo einszweidrei | cryptsetup luksFormat --pbkdf-memory 32768 {disk}")
+        m.execute(f"echo einszweidrei | cryptsetup luksOpen {disk} dm-test")
+        m.execute("mkfs.ext4 /dev/mapper/dm-test; mount /dev/mapper/dm-test /mnt; echo Hi >/mnt/hello; umount /mnt")
+        m.execute("cryptsetup close dm-test")
```
A test for a Filesystem with only directories would be nice as well.

mvollmer replied:
Yes, agreed.

mvollmer replied:
Done.

garrett commented:
Are the screenshots up to date? They still vary wildly from the mockups in several ways.

mvollmer commented on pkg/storaged/dialog.jsx:
```diff
@@ -1269,24 +1339,51 @@ export function teardown_danger_message(usage, expect_single_unmount) {
+                                                           [decode_filename(u.block.PreferredDevice)],
+                                                           { superuser: true, err: "message" });
+                        if (empty.trim() != "yes") {
+                            console.log("INFO", empty);
```
Let's just remove this!

cockpituous commented on pkg/storaged/dialog.jsx:
```diff
@@ -1269,24 +1339,54 @@ export function teardown_danger_message(usage, expect_single_unmount) {
+                                u.data_warning = cockpit.format(_("$0 used, $1 total"),
+                                                                fmt_size((info.total - info.free) * info.unit),
+                                                                fmt_size(info.total * info.unit));
+                            } catch {
+                                u.data_warning = _("Device contains unrecognized data");
```
These 2 added lines are not executed by any test.

cockpituous replied:
These 2 added lines are not executed by any test.

cockpituous replied:
These 2 added lines are not executed by any test.

cockpituous replied:
These 2 added lines are not executed by any test.

mvollmer commented:
This PR causes pixel test failures in Anaconda. Not sure how we handle this situation...

mvollmer commented:
> This PR causes pixel test failures in Anaconda. Not sure how we handle this situation...

I wasn't able to push the new pixel references to rhinstaller/anaconda-webui.

jelly commented:
> > This PR causes pixel test failures in Anaconda. Not sure how we handle this situation...
> 
> I wasn't able to push the new pixel references to rhinstaller/anaconda-webui.

Oh, that sounds like a permission issue? I can probably push that as I am still in the rhinstaller team.

mvollmer commented on pkg/storaged/fsys-is-empty.sh:
```diff
@@ -0,0 +1,34 @@
+    fi
+}
+
+trap maybe_unmount 0
```
Fine with me.

cockpituous commented on pkg/storaged/dialog.jsx:
```diff
@@ -1269,24 +1339,54 @@ export function teardown_danger_message(usage, expect_single_unmount) {
+                                u.data_warning = cockpit.format(_("$0 used, $1 total"),
+                                                                fmt_size((info.total - info.free) * info.unit),
+                                                                fmt_size(info.total * info.unit));
+                            } catch {
+                                u.data_warning = _("Device contains unrecognized data");
```
These 2 added lines are not executed by any test.

garrett commented:
What happened to all the changes I detailed that should happen?

This shouldn't have been merged. I don't see any follow-up either.
